src/dune_engine/build_system.ml

```diff
@@ -64,14 +71,26 @@ end = struct
       db := Path.Set.add !db p
     )
 
-  let load () = Option.value ~default:Path.Set.empty (P.load fn)
+  let load =
+    let loaded = ref None in
```

Comment:
This is probably not the final state of the code because as it stands `remove_now` doesn't really work, but let me clarify the design I was thinking of.

Instead of having `let db = ref Path.Set.empty` on line 62 we can start by initializing that mutable ref from the file, with something like this:

```
let db = lazy (ref (Option.value ~default:Path.Set.empty (P.load fn)))
``` 

Then you keep mutating this ref (and setting needs_dumping to true) whenever you make the corresponding filesystem modifications.

We won't need the `load` function anymore.
In `dump` function we won't need the call to `union`, anymore.
We shouldn't have to call `P.dump` from `remove_now`, either, because `needs_dumping` should take care of that.